### ISSUE TITLE
`no-array-fill-with-reference-type`: false positive for typed custom `.fill()` receiver returned by a function

### ISSUE THREAD

**Issue by Sli4o** [sindresorhus/eslint-plugin-unicorn#3689]
The `unicorn/no-array-fill-with-reference-type` rule reports a user-defined `.fill()` method when the receiver comes from a function call and the argument is a `const` initialized with an object literal.

This is a reduced example; the actual custom type and implementation are unrelated to arrays:

```ts
interface CustomTemplate {
	fill(values: Record<string, string>): unknown;
}

declare function restoreTemplate(): CustomTemplate;

const restored = restoreTemplate();
const values = {title: 'World'};

// Reported: "Do not use a reference value as the fill value."
restored.fill(values);
```

Rule: `unicorn/no-array-fill-with-reference-type`

Expected: No report. TypeScript knows that `restored` is a `CustomTemplate`, whose `fill()` method is user-defined and is not `Array#fill()`.

Actual: The argument `values` is reported as a reference fill value because the receiver is syntactically unknown after the function call.

Notably, equivalent calls are not reported when the receiver is syntactically recognizable as a non-array, for example an instance created with `new CustomTemplate()` or an object literal with its own `fill()` method. In the affected project, passing the object literal directly also does not produce the report, while assigning it to a `const` first does.

The false positive appears to result from treating an unknown receiver as a possible array without using the receiver's TypeScript type information.

**Comment by Sli4o** [comment]
Thanks for fixing the original reproducer. I tested this again with the current version: the original reproducer no longer reports, but the same false positive still occurs when the typed custom `.fill()` receiver is accessed through nested properties.

Updated reproducer:

```ts
interface CustomTemplate {
	fill(values: Record<string, string>): unknown;
}

interface Context {
	templates: {default: CustomTemplate};
}

declare const context: Context;

// Reported: "Do not use a reference value as the fill value."
context.templates.default.fill({data: 'xyz'});
```

Expected: No report. TypeScript knows that `context.templates.default` is a `CustomTemplate`, and its `fill()` method is user-defined, not `Array#fill()`.

So the previous fix appears to handle the direct typed receiver case (`restored.fill(...)`), but not a typed receiver reached through a member-expression chain (`context.templates.default.fill(...)`).

**Comment by sindresorhus** [comment]
This works with type-aware linting. Without it, the rule would need to resolve property types through interfaces, aliases, generics, etc., which is basically reimplementing the TypeScript type checker. That is out of scope, so unknown receivers are still reported, as documented.

Enable type-aware linting, or annotate the receiver: `const template: CustomTemplate = context.templates.default;`

**Comment by Sli4o** [comment]
Thanks for the clarification! You were right: I wasn't using "type-aware linting." Switching from `recommended` to `recommendedTypeChecked` in typescript-eslint fixed the false positive in my actual project as well.

Sorry for the confusion, and thanks for the helpful tip!
